src/services/utilities.ts

```diff
@@ -3273,23 +3273,61 @@ namespace ts {
+        let containedNodeModulesWatchers: ESMap<string, FileWatcher> | undefined;
+        let cache: ESMap<Path, boolean | { modulePaths: readonly ModulePath[], moduleSpecifiers?: readonly string[] }> | undefined;
         let importingFileName: Path | undefined;
         const wrapped: ModuleSpecifierCache = {
             get(fromFileName, toFileName) {
                 if (!cache || fromFileName !== importingFileName) return undefined;
                 return cache.get(toFileName);
             },
-            set(fromFileName, toFileName, moduleSpecifiers) {
+            set(fromFileName, toFileName, cached) {
+                if (cache && fromFileName !== importingFileName) {
+                    this.clear();
+                }
+                importingFileName = fromFileName;
+                (cache ||= new Map()).set(toFileName, cached);
+
+                // If any module specifiers were generated based off paths in node_modules,
+                // a package.json file in that package was read and is an input to the cached.
+                // Instead of watching each individual package.json file, set up a wildcard
+                // directory watcher for any node_modules referenced and clear the cache when
+                // it sees any changes.
+                if (cached.moduleSpecifiers) {
+                    for (const p of cached.modulePaths) {
+                        if (p.isInNodeModules) {
+                            const nodeModulesPath = p.path.substring(0, p.path.indexOf(nodeModulesPathPart) + nodeModulesPathPart.length);
+                            if (!containedNodeModulesWatchers?.has(nodeModulesPath)) {
+                                (containedNodeModulesWatchers ||= new Map()).set(
+                                    nodeModulesPath,
+                                    host.watchNodeModulesDirectory(nodeModulesPath, this.clear),
+                                );
+                            }
+                        }
+                    }
+                }
+            },
+            getModulePaths(fromFileName, toFileName) {
+                if (!cache || fromFileName !== importingFileName) return undefined;
+                const cached = cache.get(toFileName);
+                return typeof cached === "object" ? cached.modulePaths : cached;
+            },
+            setModulePaths(fromFileName, toFileName, modulePaths) {
                 if (cache && fromFileName !== importingFileName) {
-                    cache.clear();
+                    this.clear();
                 }
                 importingFileName = fromFileName;
-                (cache ||= new Map()).set(toFileName, moduleSpecifiers);
+                (cache ||= new Map()).set(toFileName, typeof modulePaths === "boolean" ? modulePaths : { modulePaths });
             },
             clear() {
-                cache = undefined;
+                containedNodeModulesWatchers?.forEach(watcher => watcher.close());

```

Comment:
Little concerned about adding and removing watches here.. since this cache is cleared between the open files it seems like we may need to create and remove and again create same kinds of watches over and over.. if there are no watches for that directory .. eg there are no closed script infos for that node_modules folder in any program or something, this may be costly on linux where every recursive directory watching needs to go traverse the directory and create watches since there is no native recurisve watching support...

Reply:
Hm, that makes sense. In most cases this will only find one node_modules directory and it will be watched pretty much every time the cache is created. In some monorepo styles, there could commonly be two node_modules directories visible to any given file—one for the subproject and one at the monorepo root. In those situations, it seems like we would eventually want to clean up old watches—what do you think is a good point in the project lifecycle to do that?

Reply:
can you think of a case where closed script info is not used but we are watching for package.json in the node_modules directory?

Reply:
Yes .. things in program will watch the node modules folder so we are good there